src/Workspaces/CSharp/Portable/LanguageServices/CSharpSemanticFactsService.cs

```diff
@@ -227,6 +227,38 @@ public ForEachSymbols GetForEachSymbols(SemanticModel semanticModel, SyntaxNode
             }
         }
 
+        public ImmutableArray<IMethodSymbol> GetDeconstructionSymbols(SemanticModel semanticModel, SyntaxNode node)
+        {
+            if (node is AssignmentExpressionSyntax assignment && assignment.IsDeconstruction())
+            {
+                return getMethods(semanticModel.GetDeconstructionInfo(assignment)).ToImmutableArray();
+            }
+
+            if (node is ForEachVariableStatementSyntax @foreach)
+            {
+                return getMethods(semanticModel.GetDeconstructionInfo(@foreach)).ToImmutableArray();
+            }
+
+            return ImmutableArray<IMethodSymbol>.Empty;
+
+            IEnumerable<IMethodSymbol> getMethods(DeconstructionInfo deconstruction)
+            {
+                var method = deconstruction.Method;
+                if (method != null)
+                {
+                    yield return method;
+                }
+
+                foreach (var nested in deconstruction.Nested)
+                {
+                    foreach (var m in getMethods(nested))
```

Comment:
:memo: this is a bit unfortunate, but I don't expect deeply nested deconstructions.

Reply:
I'd prefer an  immutable array builder.  we end up tending to have to avoid IEnumreables in pathological cases.